core/lib/vm_utils/src/vm_env.rs

```diff
@@ -0,0 +1,185 @@
+use multivm::{
+    interface::{L1BatchEnv, L2BlockEnv, SystemEnv, TxExecutionMode},
+    vm_latest::constants::BLOCK_GAS_LIMIT,
+};
+use zksync_contracts::BaseSystemContracts;
+use zksync_dal::StorageProcessor;
+use zksync_types::{
+    fee_model::BatchFeeInput, Address, L1BatchNumber, L2ChainId, MiniblockNumber,
+    ProtocolVersionId, H256, ZKPORTER_IS_AVAILABLE,
+};
+
+/// The environment required to initialize the VM.
+#[derive(Clone, Debug)]
+pub struct VmEnv {
+    pub l1_batch_env: L1BatchEnv,
+    pub system_env: SystemEnv,
+}
+
+/// The builder for the `VmEnv`. It allows to load the required data from the database.
+#[derive(Clone, Debug)]
+pub struct VmEnvBuilder {
+    miniblock_number: Option<MiniblockNumber>,
+    l1_batch_number: L1BatchNumber,
+    prev_batch_hash: Option<H256>,
+    fee_account: Option<Address>,
+    validation_computational_gas_limit: u32,
+    chain_id: L2ChainId,
+}
+
+impl VmEnvBuilder {
+    pub fn new(
+        l1_batch_number: L1BatchNumber,
+        validation_computational_gas_limit: u32,
+        chain_id: L2ChainId,
+    ) -> Self {
+        Self {
+            l1_batch_number,
+            validation_computational_gas_limit,
+            chain_id,
+            miniblock_number: None,
+            prev_batch_hash: None,
+            fee_account: None,
+        }
+    }
+    pub fn with_miniblock_number(mut self, miniblock_number: MiniblockNumber) -> Self {
```

Comment:
Same problem as with `with_prev_batch_hash` here; unnecessarily brittle. In this particular case, you could access `miniblock_number ` from `L1BatchEnv` in `create_vm_for_l1_batch` (IIUC, that's the only use of the method).

Reply:
It will be more use, where i need to have access to the different miniblock.

Reply:
but at the same time we can use this method for creating wrong data intentionaly.